### PR TITLE
CloudSearch Domain Resource Support

### LINKED ISSUE #7833: Support for Managing Amazon CloudSearch Domains
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

This feature request is to gauge interest in supporting Terraform resources for [Amazon CloudSearch](https://aws.amazon.com/cloudsearch/).

### New or Affected Resource(s)

* `aws_cloudsearch_domain`
* Potentially others as determined by use cases

### Potential Terraform Configuration

```hcl
# This is a stub Terraform Configuration for the feature request
# and does not contain full details.
resource "aws_cloudsearch_domain" "example" {
  name = "example"
}
```

### References

* Previous implementation attempt: https://github.com/terraform-providers/terraform-provider-aws/pull/1541
* AWS Go SDK Reference: https://docs.aws.amazon.com/sdk-for-go/api/service/cloudsearch/
* API Reference: https://docs.aws.amazon.com/cloudsearch/latest/developerguide/api-ref.html

### PR DESCRIPTION
**Pull request by celestialorb**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #7833

Output from acceptance testing:
```
make testacc TESTARGS="-run TestAccAWSCloudSearchDomain"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run TestAccAWSCloudSearchDomain -timeout 120m
=== RUN   TestAccAWSCloudSearchDomain_basic
=== PAUSE TestAccAWSCloudSearchDomain_basic
=== RUN   TestAccAWSCloudSearchDomain_textAnalysisScheme
=== PAUSE TestAccAWSCloudSearchDomain_textAnalysisScheme
=== RUN   TestAccAWSCloudSearchDomain_badName
=== PAUSE TestAccAWSCloudSearchDomain_badName
=== RUN   TestAccAWSCloudSearchDomain_badInstanceType
=== PAUSE TestAccAWSCloudSearchDomain_badInstanceType
=== RUN   TestAccAWSCloudSearchDomain_badIndexFieldNames
=== PAUSE TestAccAWSCloudSearchDomain_badIndexFieldNames
=== RUN   TestAccAWSCloudSearchDomain_badIndexFieldType
=== PAUSE TestAccAWSCloudSearchDomain_badIndexFieldType
=== CONT  TestAccAWSCloudSearchDomain_basic
=== CONT  TestAccAWSCloudSearchDomain_badIndexFieldNames
=== CONT  TestAccAWSCloudSearchDomain_badIndexFieldType
=== CONT  TestAccAWSCloudSearchDomain_badName
=== CONT  TestAccAWSCloudSearchDomain_badInstanceType
=== CONT  TestAccAWSCloudSearchDomain_textAnalysisScheme
--- PASS: TestAccAWSCloudSearchDomain_badInstanceType (2.85s)
--- PASS: TestAccAWSCloudSearchDomain_badIndexFieldType (3.01s)
--- PASS: TestAccAWSCloudSearchDomain_badName (3.04s)
--- PASS: TestAccAWSCloudSearchDomain_badIndexFieldNames (4.18s)
--- PASS: TestAccAWSCloudSearchDomain_textAnalysisScheme (41.59s)
--- PASS: TestAccAWSCloudSearchDomain_basic (60.54s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       60.616s
```

### CODE REVIEW

**Comment by celestialorb** [comment]
Only question I have at the moment from someone more intimately familiar with Terraform resource/provider design, is should an index in a CloudSearch domain be a separate resource? I feel as if it should be due to it's complexity, but perhaps the initial approach taken by @Mi3-14159 is perfectly fine.

**Comment by austinkelleher** [comment]
Nice work @celestialorb! This would be extremely helpful to our team! Is there something blocking the review of this PR?

**Comment by celestialorb** [comment]
@austinkelleher Not sure, haven't seen anyone official pop into the PR yet. Don't know what can be done to get it noticed; I'm sure there's a ton of PRs to review on their end.

In the meantime I've published this custom version of the provider under my own namespace, `celestialorb/aws`. If you use that as your source in your Terraform it should pull this version and you'll be able to use it for CloudSearch domains.

File: website/docs/r/cloudsearch_domain.html.markdown [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+# Resource: aws_cloudsearch_domain
+
+Provides an CloudSearch domain resource.
+
+## Example Usage
+
+```terraform
+resource "aws_cloudsearch_domain" "my_domain" {
+  domain_name   = "test-domain"
```
Comment (hallsamuel90) [inline website/docs/r/cloudsearch_domain.html.markdown:17]: I think this should be `name` instead of `domain_name`

Comment (celestialorb) [reply website/docs/r/cloudsearch_domain.html.markdown]: Thanks, good catch!

File: aws/resource_aws_cloudsearch_domain.go [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+		DomainNames: []*string{
+			aws.String(d.Get("name").(string)),
+		},
+	}
+
+	resp, err := conn.DescribeDomains(&domainlist)
+	if err != nil {
+		return err
+	}
+	domain := resp.DomainStatusList[0]
```
Comment (hallsamuel90) [inline aws/resource_aws_cloudsearch_domain.go:188]: this is throwing an index out-of-range panic error here when running `plan`. Any ideas to the cause would be super helpful! I'm using one of the sample configs from the acceptance tests for some additional context.

`panic: runtime error: index out of range [0] with length 0`

Comment (celestialorb) [reply aws/resource_aws_cloudsearch_domain.go:188]: I'd check the name of your domain and ensure that it matches the one you're expecting in your AWS account. From this it just seems like it wasn't able to find the domain from the given name where it is expecting one to be. Code could definitely be improved upon here with some additional sanity checks.

**Comment by hallsamuel90** [comment]
Awesome job @celestialorb !! It's working great and couldn't have come at a better time for our team!

**Comment by vsharma29** [comment]
@celestialorb Thanks for the provider. I am getting an issue even if i dont change the index elements still the plan shows me changes if i do terraform plan and it recreates all of the index everytime.


resource "aws_cloudsearch_domain" "domain" {
  name = var.domain_name
  instance_type = var.instance_type
  multi_az = true

  dynamic "index" {
    for_each = var.indexes
    content {
      name = index.value["name"]
      type = index.value["type"]
      search = lookup(index, "search", true)
      facet  = lookup(index, "facet", true)
      return = lookup(index, "return", true)
      sort = lookup(index, "sort", true)
      highlight = lookup(index, "highlight", true)
      analysis_scheme = lookup(index, "analysis_scheme", "_en_default_")
    }
  }
  tags = var.tags

  service_access_policies = data.aws_iam_policy_document.cloudsearch_access_policy.json
}

Value from vars file

indexes = [
  {
    name = "abstract"
    type = "text"
  },
  {
    name = "analystname"
    type = "text-array"
  },
  {
    name = "htmtext"
    type = "text"
  },
  {
    name = "issuername"
    type = "text"
  },
  {
    name = "organizationid"
    type = "text"
  },
  {
    name = "pdftext"
    type = "text"
  },
  {
    name = "reporttype"
    type = "text"
  },
  {
    name = "sectorlevel1"
    type = "text"
  },
  {
    name = "sectorlevel2"
    type = "text"
  },
  {
    name = "subtitle"
    type = "text"
  },
  {
    name = "ticker"
    type = "text"
  },
  {
    name = "title"
    type = "text"
  },
  {
    name = "url"
    type = "text"
  }
]

**Comment by celestialorb** [comment]
> @celestialorb Thanks for the provider. I am getting an issue even if i dont change the index elements still the plan shows me changes if i do terraform plan and it recreates all of the index everytime.
> 
> ```terraform
> resource "aws_cloudsearch_domain" "domain" {
>   name = var.domain_name
>   instance_type = var.instance_type
>   multi_az = true
> 
>   dynamic "index" {
>     for_each = var.indexes
>     content {
>       name = index.value["name"]
>       type = index.value["type"]
>       search = lookup(index, "search", true)
>       facet = lookup(index, "facet", true)
>       return = lookup(index, "return", true)
>       sort = lookup(index, "sort", true)
>       highlight = lookup(index, "highlight", true)
>       analysis_scheme = lookup(index, "analysis_scheme", "_en_default_")
>     }
>   }
>
>   tags = var.tags
> 
>   service_access_policies = data.aws_iam_policy_document.cloudsearch_access_policy.json
> }
> 
> Value from vars file
> 
> indexes = [
> {
>   name = "abstract"
>   type = "text"
> },
> {
>   name = "analystname"
>   type = "text-array"
> },
> {
>   name = "htmtext"
>   type = "text"
> },
> {
>   name = "issuername"
>   type = "text"
> },
> {
>   name = "organizationid"
>   type = "text"
> },
> {
>   name = "pdftext"
>   type = "text"
> },
> {
>   name = "reporttype"
>   type = "text"
> },
> {
>   name = "sectorlevel1"
>   type = "text"
> },
> {
>   name = "sectorlevel2"
>   type = "text"
> },
> {
>   name = "subtitle"
>   type = "text"
> },
> {
>   name = "ticker"
>   type = "text"
> },
> {
>   name = "title"
>   type = "text"
> },
> {
>   name = "url"
>   type = "text"
> }
> ]
> ```

The issue here is that you're setting an `analysis_scheme` for indices where one doesn't make sense (and can't be set). Analysis schemes are only valid for `text` and `text-array` fields, but in your Terraform configuration you're setting the `analysis_scheme` for all indices via the `lookup` function:

```
analysis_scheme = lookup(index, "analysis_scheme", "_en_default_")
```

There's still some work to done surrounding indices, given that they aren't exactly straight-forward and I think the approach originally taken can be improved upon.

**Comment by vsharma29** [comment]
@celestialorb All the fields which i have set up are text or text array so it can be set . I am not sure whats driving the plan to
Misbehave but something wierd going on

**Comment by celestialorb** [comment]
> @celestialorb All the fields which i have set up are text or text array so it can be set . I am not sure whats driving the plan to
> Misbehave but something wierd going on

My mistake, I thought I had spotted in `int` in there somewhere. I was able to reproduce the issue earlier when using non-text fields like `int`, but not with only `text` fields. I'll go back and take another look later.

**Comment by celestialorb** [comment]
> @celestialorb All the fields which i have set up are text or text array so it can be set . I am not sure whats driving the plan to
> Misbehave but something wierd going on

Including a copy of the plan would help to debug the issue.

**Comment by celestialorb** [comment]
> @celestialorb All the fields which i have set up are text or text array so it can be set . I am not sure whats driving the plan to
> Misbehave but something wierd going on

Yeah, I'll need to see what your plan looks like. I can't reproduce the issue on my end.

**Comment by vsharma29** [comment]
Sorry about late reply. Copy of my plan.`[](url)
[pipelineLog-10.txt](https://github.com/hashicorp/terraform-provider-aws/files/6576711/pipelineLog-10.txt)

**Comment by celestialorb** [comment]
> Sorry about late reply. Copy of my plan.`
> [pipelineLog-10.txt](https://github.com/hashicorp/terraform-provider-aws/files/6576711/pipelineLog-10.txt)

Looks like it's because you're changing the `search` attribute from `false` to `true` on a lot of these indices, it not all, so it's recreating them. This is expected behavior.

**Comment by YakDriver** [comment]
@celestialorb I recommend setting this to allow maintainers to edit so that we can help you update this PR for service packages (see #20000). If you're no longer involved in this sphere, we can also cherry pick your commits to a new branch.

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

##### Commercial

```console
% make testacc TESTARGS='-run=TestAccCloudSearchDomain_' PKG=cloudsearch 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudsearch/... -v -count 1 -parallel 20 -run=TestAccCloudSearchDomain_ -timeout 180m
=== RUN   TestAccCloudSearchDomain_basic
=== PAUSE TestAccCloudSearchDomain_basic
=== RUN   TestAccCloudSearchDomain_disappears
=== PAUSE TestAccCloudSearchDomain_disappears
=== RUN   TestAccCloudSearchDomain_indexFields
=== PAUSE TestAccCloudSearchDomain_indexFields
=== RUN   TestAccCloudSearchDomain_update
=== PAUSE TestAccCloudSearchDomain_update
=== CONT  TestAccCloudSearchDomain_basic
=== CONT  TestAccCloudSearchDomain_indexFields
=== CONT  TestAccCloudSearchDomain_disappears
=== CONT  TestAccCloudSearchDomain_update
--- PASS: TestAccCloudSearchDomain_disappears (662.90s)
--- PASS: TestAccCloudSearchDomain_basic (679.61s)
--- PASS: TestAccCloudSearchDomain_indexFields (1588.94s)
--- PASS: TestAccCloudSearchDomain_update (1789.34s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/cloudsearch        1792.704s
% make testacc TESTARGS='-run=TestAccCloudSearchDomainServiceAccessPolicy_' PKG=cloudsearch
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudsearch/... -v -count 1 -parallel 20 -run=TestAccCloudSearchDomainServiceAccessPolicy_ -timeout 180m
=== RUN   TestAccCloudSearchDomainServiceAccessPolicy_basic
=== PAUSE TestAccCloudSearchDomainServiceAccessPolicy_basic
=== RUN   TestAccCloudSearchDomainServiceAccessPolicy_update
=== PAUSE TestAccCloudSearchDomainServiceAccessPolicy_update
=== CONT  TestAccCloudSearchDomainServiceAccessPolicy_basic
=== CONT  TestAccCloudSearchDomainServiceAccessPolicy_update
--- PASS: TestAccCloudSearchDomainServiceAccessPolicy_basic (1807.86s)
--- PASS: TestAccCloudSearchDomainServiceAccessPolicy_update (2546.93s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/cloudsearch        2550.273s
```

##### GovCloud

```console
% make testacc TESTARGS='-run=TestAccCloudSearchDomain_' PKG=cloudsearch
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudsearch/... -v -count 1 -parallel 20 -run=TestAccCloudSearchDomain_ -timeout 180m
=== RUN   TestAccCloudSearchDomain_basic
=== PAUSE TestAccCloudSearchDomain_basic
=== RUN   TestAccCloudSearchDomain_disappears
=== PAUSE TestAccCloudSearchDomain_disappears
=== RUN   TestAccCloudSearchDomain_indexFields
=== PAUSE TestAccCloudSearchDomain_indexFields
=== RUN   TestAccCloudSearchDomain_update
=== PAUSE TestAccCloudSearchDomain_update
=== CONT  TestAccCloudSearchDomain_basic
=== CONT  TestAccCloudSearchDomain_indexFields
=== CONT  TestAccCloudSearchDomain_disappears
=== CONT  TestAccCloudSearchDomain_update
=== CONT  TestAccCloudSearchDomain_basic
    acctest.go:608: skipping tests; partition aws-us-gov does not support cloudsearch service
=== CONT  TestAccCloudSearchDomain_indexFields
    acctest.go:608: skipping tests; partition aws-us-gov does not support cloudsearch service
--- SKIP: TestAccCloudSearchDomain_indexFields (1.47s)
--- SKIP: TestAccCloudSearchDomain_basic (1.47s)
=== CONT  TestAccCloudSearchDomain_disappears
    acctest.go:608: skipping tests; partition aws-us-gov does not support cloudsearch service
--- SKIP: TestAccCloudSearchDomain_disappears (1.47s)
=== CONT  TestAccCloudSearchDomain_update
    acctest.go:608: skipping tests; partition aws-us-gov does not support cloudsearch service
--- SKIP: TestAccCloudSearchDomain_update (1.47s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/cloudsearch        4.968s
% make testacc TESTARGS='-run=TestAccCloudSearchDomainServiceAccessPolicy_' PKG=cloudsearch
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/cloudsearch/... -v -count 1 -parallel 20 -run=TestAccCloudSearchDomainServiceAccessPolicy_ -timeout 180m
=== RUN   TestAccCloudSearchDomainServiceAccessPolicy_basic
=== PAUSE TestAccCloudSearchDomainServiceAccessPolicy_basic
=== RUN   TestAccCloudSearchDomainServiceAccessPolicy_update
=== PAUSE TestAccCloudSearchDomainServiceAccessPolicy_update
=== CONT  TestAccCloudSearchDomainServiceAccessPolicy_basic
=== CONT  TestAccCloudSearchDomainServiceAccessPolicy_update
=== CONT  TestAccCloudSearchDomainServiceAccessPolicy_basic
    acctest.go:608: skipping tests; partition aws-us-gov does not support cloudsearch service
--- SKIP: TestAccCloudSearchDomainServiceAccessPolicy_basic (1.30s)
=== CONT  TestAccCloudSearchDomainServiceAccessPolicy_update
    acctest.go:608: skipping tests; partition aws-us-gov does not support cloudsearch service
--- SKIP: TestAccCloudSearchDomainServiceAccessPolicy_update (1.30s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/cloudsearch        4.651s
```

**Pull request merged**